plutus-tx-plugin/test/Plugin/Data/Spec.hs

```diff
@@ -1,20 +1,70 @@
-{-# LANGUAGE TypeApplications    #-}
-{-# LANGUAGE TypeFamilies        #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# OPTIONS_GHC -Wno-incomplete-patterns #-}
+-- pattern match to avoid type getting simplified away
+{-
+This is one of the test cases that reveals a bug in GHC 9: it fails to perform some
+eta reductions that should be performed.
+
+The CoreExpr for this test case in GHC 8 is simply @Mono1@, but in GHC 9, it becomes
+@\ds1 ds2 -> Mono1 ds1 ds2@. This leads to bigger PIR, since we don't do eta reduction in PIR.
+
+The reason for the GHC 9 behavior appears to be this Multiplicity check:
+https://gitlab.haskell.org/ghc/ghc/-/blob/a54827e0b48af33fa9cfde6ad131c6751c2fe321/compiler/GHC/Core/Utils.hs#L2501.
+GHC thinks @Mono1@ has a linear type, but @ds1@ and @ds2@ are non-linear, and the check is False.
+
+The GHC 8 behavior is restored if @MyMonoData@ is defined this way:
+
+@
+data MyMonoData where
+  Mono1 :: Integer %Many -> Integer %Many -> MyMonoData
+  Mono2 :: Integer %Many -> MyMonoData
+  Mono3 :: Integer %Many -> MyMonoData
+@
+-}
+-- Bang patterns on pattern-matches do nothing: it's already strict
+-- pattern match to avoid type getting simplified away
+-- pattern match to avoid type getting simplified away
+-- checks that the type of 'strictField2' is replaced with 'Integer', see Note [On data constructor workers and wrappers]
+-- checks that the 'wrapper' is compiled but unused, see Note [On data constructor workers and wrappers]
+-- must be compiled with a lazy case
+-- Bang patterns on data types do nothing: fields are already strict
+-- pattern match to avoid type getting simplified away
+-- pattern match to avoid type getting simplified away
+-- This will generate code using 'build' if we're on greater than -O0. That's not optimal for
+-- us, since we don't have any rewrite rules to fire, but it's fine and we can handle it.
+-- It is very difficult to get GHC to make a non-polymorphic redex if you use
+-- list literal syntax with integers. But this works.
+{- Note [Non-regular data types in tests]
+A non-regular data type, a.k.a. a nested data type is, quoting "Nested Datatypes" by Richard Bird
+and Lambert Meertens, a parametrised data type whose declaration involves different instances of the
+accompanying type parameters. Such data types cannot be encoded with regular @fix :: (* -> *) -> *@
+and require our fancy @ifix@ business, hence they make for good tests, which is how we have so many
+of them.
+-}
+-- See Note [Non-regular data types in tests].
+-- TODO: replace this with 'first' when we have working recursive functions
+-- See Note [Non-regular data types in tests].
+-- The type signatures are needed due to a bug (see 'emptyRoseNewId')
+-- See Note [Non-regular data types in tests].
+-- Note that the parameters get swapped.
+-- Despite the type family being applied to a parameterized type we can still reduce it
+-- Here we cannot reduce the type family
```

Comment:
Lol, did it just move all the comments in this module to the beginning of the module? Wonderful.

Reply:
> Lol, did it just move all the comments in this module to the beginning of the module? Wonderful.

:laughing: That's actually pretty impressive.  It's taken most (but not quite all) of the comments and put them [between two `OPTIONS_GHC` pragmas](https://github.com/IntersectMBO/plutus/blob/2dfc62146e729d24d6abe6d729cbc00468df5b5a/plutus-tx-plugin/test/Plugin/Data/Spec.hs#L9) at the top of the file!